lib/vrl/compiler/src/type_def.rs

```diff
@@ -151,21 +152,21 @@ impl KindInfo {
     /// }
     ///
     /// e.g., the existing [`KindInfo`] gets nested into the provided path.
-    pub fn for_path(mut self, path: Path) -> Self {
-        use path::Segment;
-
-        for segment in path.segments().iter().rev() {
+    pub fn for_path(mut self, path: LookupBuf) -> Self {
+        for segment in path.iter().rev() {
             match segment {
-                Segment::Field(field) => {
+                SegmentBuf::Field(FieldBuf { name, .. }) => {
                     let mut map = BTreeMap::default();
-                    map.insert(Field::Field(field.as_str().to_owned()), self);
+                    map.insert(Field::Field(name.as_str().to_owned()), self);
 
                     let mut set = BTreeSet::new();
                     set.insert(TypeKind::Object(map));
 
                     self = KindInfo::Known(set);
                 }
-                Segment::Coalesce(fields) => {
+                SegmentBuf::Coalesce(fields) => {
+                    // TODO: I'm not sure this is right - it should be the
+                    // combined typedef of all the fields in the coalesce.
```

Comment:
Do we want to add a test to validate this?

Reply:
From what I can tell, I haven't changed the actual result of the code here. But I'm not really sure what it is doing? Why is it taking the last field of the coalesce and returning an Object typekind with that fieldname? Do you have any ideas?